migrate_test.go

```diff
@@ -1414,3 +1417,282 @@ func equalDbSeq(t *testing.T, i int, expected migrationSequence, got *dStub.Stub
+
+	tests := []struct {
+		migrationFiles []string
+		targetVersion  uint
+		remainingFiles []string
+		emptyDestPath  bool
+	}{
+		{
+			migrationFiles: []string{"1_name.up.sql", "2_name.up.sql", "3_name.up.sql"},
+			targetVersion:  2,
+			remainingFiles: []string{"1_name.up.sql", "2_name.up.sql"},
+		},
+		{
+			migrationFiles: []string{"1_name.up.sql", "2_name.up.sql", "3_name.up.sql", "4_name.up.sql", "5_name.up.sql"},
+			targetVersion:  3,
+			remainingFiles: []string{"1_name.up.sql", "2_name.up.sql", "3_name.up.sql"},
+		},
+		{
+			migrationFiles: []string{},
+			targetVersion:  1,
+			remainingFiles: []string{},
+			emptyDestPath:  true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run("", func(t *testing.T) {
+			for _, file := range test.migrationFiles {
+				if err := os.WriteFile(filepath.Join(tempDir, file), []byte(""), 0644); err != nil {
+					t.Fatal(err)
+				}
+			}
+
+			if test.emptyDestPath {
+				m.dirtyStateConf.destPath = ""
+			}
+
+			if err := m.cleanupFiles(test.targetVersion); err != nil {
+				t.Fatal(err)
+			}
+
+			// check 1: only files upto the target version should exist
+			for _, file := range test.remainingFiles {
+				if _, err := os.Stat(filepath.Join(tempDir, file)); os.IsNotExist(err) {
+					t.Fatalf("expected file %s to exist, but it does not", file)
+				}
+			}
+
+			// check 2: the files removed are as expected
+			deletedFiles := diff(test.migrationFiles, test.remainingFiles)
+			for _, deletedFile := range deletedFiles {
+				if _, err := os.Stat(filepath.Join(tempDir, deletedFile)); !os.IsNotExist(err) {
+					t.Fatalf("expected file %s to be deleted, but it still exists", deletedFile)
+				}
+			}
+		})
+	}
+}
+
+func TestCopyFiles(t *testing.T) {
+	srcDir, cleanupSrc := setupTempDir(t)
+	defer cleanupSrc()
+
+	destDir, cleanupDest := setupTempDir(t)
+	defer cleanupDest()
+
+	m, _ := setupMigrateInstance(destDir)
+	m.dirtyStateConf.srcPath = srcDir
+
+	tests := []struct {
+		migrationFiles []string
+		copiedFiles    []string
+		emptyDestPath  bool
+	}{
+		{
+			migrationFiles: []string{"1_name.up.sql", "2_name.up.sql", "3_name.up.sql"},
+			copiedFiles:    []string{"1_name.up.sql", "2_name.up.sql", "3_name.up.sql"},
+		},
+		{
+			migrationFiles: []string{"1_name.up.sql", "2_name.up.sql", "3_name.up.sql", "4_name.up.sql", "current.sql"},
+			copiedFiles:    []string{"1_name.up.sql", "2_name.up.sql", "3_name.up.sql", "4_name.up.sql"},
+		},
+		{
+			emptyDestPath: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run("", func(t *testing.T) {
+			for _, file := range test.migrationFiles {
+				if err := os.WriteFile(filepath.Join(srcDir, file), []byte(""), 0644); err != nil {
+					t.Fatal(err)
+				}
+			}
+			if test.emptyDestPath {
+				m.dirtyStateConf.destPath = ""
+			}
+
+			if err := m.copyFiles(); err != nil {
+				t.Fatal(err)
+			}
+
+			for _, file := range test.copiedFiles {
+				if _, err := os.Stat(filepath.Join(destDir, file)); os.IsNotExist(err) {
```

Comment:
Ideally you want to check that `"current.sql"` hasn't been copied.

I don't know if using https://pkg.go.dev/os#ReadDir you can write a helper function returning all the files in a directory you can compare with an expected list of files. It would avoid in the various test-cases the two loops (one checking for existence, one checking for absence).

Reply:
I don't think it would matter. current.sql is a project specific concept, migrate file driver doesn't know anything about things that don't match its pattern.